File: exercises/javascript/1_characters_and_quantifiers/evaluate.js
import { callRegex, hissRegex, yayRegex, yeehawRegex } from ".";

// adding string boundaries so regexes behave as expected
const yeehawRegexComplete = RegExp(`^${yeehawRegex.source}$`);
const hissRegexComplete = RegExp(`^${hissRegex.source}$`);
const yayRegexComplete = RegExp(`^${yayRegex.source}$`);
const callRegexComplete = RegExp(`^${callRegex.source}$`);

describe("yeehaw regex", function () {
  it('should match yeehaw with two "e"s and no !', function () {
    expect(yeehawRegexComplete.test("yeehaw")).toBe(true);
  });
  it('should match yeehaw with four "e"s and one !', function () {
    expect(yeehawRegexComplete.test("yeeeehaw!")).toBe(true);
  });
  it("should not match yeehaw with one e", function () {
    expect(yeehawRegexComplete.test("yehaw")).toBe(false);
  });
  it('should not match yeehaw with two "e"s and four !', function () {
    expect(yeehawRegexComplete.test("yeehaw!!!!")).toBe(false);
  });
});

describe("hiss regex", function () {
  it('should match hiss with two "s"s', function () {
    expect(hissRegexComplete.test("hiss")).toBe(true);
  });
  it('should match hiss with five "s"s', function () {
    expect(hissRegexComplete.test("hisssss")).toBe(true);
  });
  it('should not match hiss with one "s"', function () {
    expect(hissRegexComplete.test("his")).toBe(false);
  });
});

describe("yay regex", function () {
  it("should not match Yay with two !s", function () {
    expect(yayRegexComplete.test("Yay!!")).toBe(false);
  });
  it("should match Yay with three !s", function () {
    expect(yayRegexComplete.test("Yay!!!")).toBe(true);
  });
  it("should match Yay with six !s", function () {
    expect(yayRegexComplete.test("Yay!!!!!!")).toBe(true);
  });
  it("should match Yay with seven !s", function () {
    expect(yayRegexComplete.test("Yay!!!!!!!")).toBe(false);
  });
});

describe("call regex", function () {
  it("should match Call me maybe with no ?s", function () {
    expect(callRegexComplete.test("Call me maybe")).toBe(true);
  });
  it("should match Call me maybe with one ?", function () {
    expect(callRegexComplete.test("Call me maybe?")).toBe(true);
  });
  it("should match Call me maybe with twelve ?s", function () {
    expect(callRegexComplete.test("Call me maybe????????????")).toBe(true);
  });
});

File: exercises/python/1_characters_and_quantifiers/evaluate.py
from unittest import TestCase
import re

from characters_regex import call_regex
from characters_regex import hiss_regex
from characters_regex import yay_regex
from characters_regex import yeehaw_regex


class EvaluateYeehaw(TestCase):

    def test_two_es_no_exclamation(self):
        matches = re.fullmatch(yeehaw_regex, 'yeehaw')
        self.assertIsNotNone(
            matches,
            'yeehaw_regex did not match "yeehaw" when it should have.'
        )

    def test_four_es_one_exclamation(self):
        matches = re.fullmatch(yeehaw_regex, 'yeeeehaw!')
        self.assertIsNotNone(
            matches,
            'yeehaw_regex did not match "yeeeehaw" when it should have.'
        )

    def test_one_e(self):
        matches = re.fullmatch(yeehaw_regex, 'yehaw')
        self.assertIsNone(
            matches,
            'yeehaw_regex matched "yehaw" when it should not have.'
        )

    def test_four_excalamtions(self):
        matches = re.fullmatch(yeehaw_regex, 'yeehaw!!!!')
        self.assertIsNone(
            matches,
            'yeehaw_regex matched "yeehaw!!!!" when it should not have.'
        )


class EvaluateHiss(TestCase):

    def test_two_s(self):
        matches = re.fullmatch(hiss_regex, 'hiss')
        self.assertIsNotNone(
            matches,
            'hiss_regex did not match "hiss" when it should have.'
        )

    def test_five_s(self):
        matches = re.fullmatch(hiss_regex, 'hisssss')
        self.assertIsNotNone(
            matches,
            'hiss_regex did not match "hisssss" when it should have.'
        )

    def test_one_s(self):
        matches = re.fullmatch(hiss_regex, 'his')
        self.assertIsNone(
            matches,
            'hiss_regex matched "his" when it should not have.'
        )


class EvaluateYay(TestCase):

    def test_two_bang(self):
        matches = re.fullmatch(yay_regex, 'Yay!!')
        self.assertIsNone(
            matches,
            'yay_regex did not match "Yay!!" when it should have.'
        )

    def test_three_bang(self):
        matches = re.fullmatch(yay_regex, 'Yay!!!')
        self.assertIsNotNone(
            matches,
            'yay_regex did not match "Yay!!!" when it should have.'
        )

    def test_six_bang(self):
        matches = re.fullmatch(yay_regex, 'Yay!!!!!!')
        self.assertIsNotNone(
            matches,
            'yay_regex did not match "Yay!!!!!!" when it should have.'
        )

    def test_seven_bang(self):
        matches = re.fullmatch(yay_regex, 'Yay!!!!!!!')
        self.assertIsNone(
            matches,
            'yay_regex matched "Yay!!!!!!!" when it should not have.'
        )


class EvaluateCall(TestCase):

    def test_no_q(self):
        matches = re.fullmatch(call_regex, 'Call me maybe')
        self.assertIsNotNone(
            matches,
            'call_regex did not match "Call me maybe" when it should have.'
        )

    def test_one_q(self):
        matches = re.fullmatch(call_regex, 'Call me maybe?')
        self.assertIsNotNone(
            matches,
            'call_regex did not match "Call me maybe?" when it should have.'
        )

    def test_twelve_q(self):
        matches = re.fullmatch(call_regex, 'Call me maybe????????????')
        self.assertIsNotNone(
            matches,
            'call_regex did not match "Call me maybe????????????" when it should have.'
        )
